## Fetch Rewards Coding Exercise - Data Analyst - Qingyang Zhou


In [289]:
# Import libraries
import numpy as np
import pandas as pd

### Loading and cleaning data

##### Receipts

In [290]:
receipts = pd.read_json('receipts.json', lines=True)
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [291]:
# # Rename userID to user_id
# receipts.rename(columns={'userID': 'user_id'}, inplace=True)

# Extract oid from dictionary, rename it to receipt_id
receipts['receipt_id'] = receipts['_id'].apply(lambda x: x['$oid'])
receipts = receipts.drop('_id', axis=1)

In [292]:
# Extract date from the 6 date columns, and convert to datetime format
def extract_convert_date(col):
    '''Extract $date from dictionary, and convert to datetime format'''
    result = col.apply(lambda x: x['$date'] if type(x) != float else np.nan)
    result = pd.to_datetime(result, unit='ms')
    return result

for col in ['createDate', 'dateScanned', 'finishedDate', 'modifyDate', 'pointsAwardedDate', 'purchaseDate']:
    receipts[col] = extract_convert_date(receipts[col])

receipts.head()

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   
1              150.0  Receipt number 5 completed, bonus point schedu...   
2                5.0                         All-receipts receipt bonus   
3                5.0                         All-receipts receipt bonus   
4                5.0                         All-receipts receipt bonus   

           createDate         dateScanned        finishedDate  \
0 2021-01-03 15:25:31 2021-01-03 15:25:31 2021-01-03 15:25:31   
1 2021-01-03 15:24:43 2021-01-03 15:24:43 2021-01-03 15:24:43   
2 2021-01-03 15:25:37 2021-01-03 15:25:37                 NaT   
3 2021-01-03 15:25:34 2021-01-03 15:25:34 2021-01-03 15:25:34   
4 2021-01-03 15:25:06 2021-01-03 15:25:06 2021-01-03 15:25:11   

           modifyDate   pointsAwardedDate  pointsEarned        purchaseDate  \
0 2021-01-03 15:25:36 2021-01-03 15:25:31         500.0 2021-01-03 00:00:00   
1 2021-01-03 15:24:48 2021-01-03 15:24:43         150.0 2021-01-02 15:24:43   
2 2021-01-03 15:25:42                 NaT           5.0 2021-01-03 00:00:00   
3 2021-01-03 15:25:39 2021-01-03 15:25:34           5.0 2021-01-03 00:00:00   
4 2021-01-03 15:25:11 2021-01-03 15:25:06           5.0 2021-01-02 15:25:06   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  \
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6   
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052   
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b   
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6   
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052   

                 receipt_id  
0  5ff1e1eb0a720f0523000575  
1  5ff1e1bb0a720f052300056b  
2  5ff1e1f10a720f052300057a  
3  5ff1e1ee0a7214ada100056f  
4  5ff1e1d20a7214ada1000561

In [293]:
# receipts.to_csv('receipts.csv')

I extract the itme list from the rewardsReceiptItemList column to a separate df. This will help design the Item table for ER model. Ideally, the list of keys will be collected from product manager/database administrator, in case the current keys are incomprehensive.

In [294]:
# Create an Item dataframe with unique keys as columns
items = [] 
for index, row in receipts.iterrows():
    if type(row['rewardsReceiptItemList']) != float:
        for item in row['rewardsReceiptItemList']:
            item['receipt_id'] = row['receipt_id']
            item['receipt_rewardsReceiptStatus'] = row['rewardsReceiptStatus']
            items.append(item)
items = pd.DataFrame(items)
items.head()


barcode                                        description finalPrice  \
0          4011                                     ITEM NOT FOUND      26.00   
1          4011                                     ITEM NOT FOUND          1   
2  028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00   
3           NaN                                                NaN        NaN   
4          4011                                     ITEM NOT FOUND      28.00   

  itemPrice needsFetchReview partnerItemId preventTargetGapPoints  \
0     26.00            False             1                   True   
1         1              NaN             1                    NaN   
2     10.00             True             2                   True   
3       NaN            False             1                   True   
4     28.00            False             1                   True   

   quantityPurchased userFlaggedBarcode userFlaggedNewItem  ... itemNumber  \
0                5.0               4011               True  ...        NaN   
1                1.0                NaN                NaN  ...        NaN   
2                1.0       028400642255               True  ...        NaN   
3                NaN               4011               True  ...        NaN   
4                4.0               4011               True  ...        NaN   

   originalMetaBriteQuantityPurchased pointsEarned targetPrice  \
0                                 NaN          NaN         NaN   
1                                 NaN          NaN         NaN   
2                                 NaN          NaN         NaN   
3                                 NaN          NaN         NaN   
4                                 NaN          NaN         NaN   

  competitiveProduct originalFinalPrice originalMetaBriteItemPrice deleted  \
0                NaN                NaN                        NaN     NaN   
1                NaN                NaN                        NaN     NaN   
2                NaN                NaN                        NaN     NaN   
3                NaN                NaN                        NaN     NaN   
4                NaN                NaN                        NaN     NaN   

  priceAfterCoupon metabriteCampaignId  
0              NaN                 NaN  
1              NaN                 NaN  
2              NaN                 NaN  
3              NaN                 NaN  
4              NaN                 NaN  

[5 rows x 36 columns]

In [295]:
# items.to_csv('items.csv')

#### Users

In [296]:
users = pd.read_json('users.json', lines=True)
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [297]:
# Extract oid from dictionary, rename it to user_id
users['user_id'] = users['_id'].apply(lambda x: x['$oid'])
users = users.drop('_id', axis=1)

In [298]:
# Extract date from the 2 date columns, and convert to datetime format
for col in ['createdDate', 'lastLogin']:
    users[col] = extract_convert_date(users[col])


In [299]:
users.head()

active             createdDate                     lastLogin      role  \
0    True 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  consumer   
1    True 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  consumer   
2    True 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  consumer   
3    True 2021-01-03 15:25:30.554 2021-01-03 15:25:30.596999936  consumer   
4    True 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  consumer   

  signUpSource state                   user_id  
0        Email    WI  5ff1e194b6a9d73a3a9f1052  
1        Email    WI  5ff1e194b6a9d73a3a9f1052  
2        Email    WI  5ff1e194b6a9d73a3a9f1052  
3        Email    WI  5ff1e1eacfcf6c399c274ae6  
4        Email    WI  5ff1e194b6a9d73a3a9f1052

#### Brands

In [300]:
brands = pd.read_json('brands.json', lines=True)
brands.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

In [301]:
# Extract oid from dictionary, rename it to brand_id
brands['brand_id'] = brands['_id'].apply(lambda x: x['$oid'])
brands = brands.drop('_id', axis=1)

In [302]:
# brands.to_csv('brands.csv')

The cpg column only contains 2 keys -- 'id' and 'ref.' The values are extracted and stored in two columns.

In [303]:
# Extract oid from dictionary, rename it to cpg_id
brands['cpg_id'] = brands['cpg'].apply(lambda x: x['$id']['$oid'])
brands['cpg_ref'] = brands['cpg'].apply(lambda x: x['$ref'])
brands = brands.drop('cpg', axis=1)
brands.head()

barcode        category      categoryCode                       name  \
0  511111019862          Baking            BAKING  test brand @1612366101024   
1  511111519928       Beverages         BEVERAGES                  Starbucks   
2  511111819905          Baking            BAKING  test brand @1612366146176   
3  511111519874          Baking            BAKING  test brand @1612366146051   
4  511111319917  Candy & Sweets  CANDY_AND_SWEETS  test brand @1612366146827   

   topBrand                      brandCode                  brand_id  \
0       0.0                            NaN  601ac115be37ce2ead437551   
1       0.0                      STARBUCKS  601c5460be37ce2ead43755f   
2       0.0  TEST BRANDCODE @1612366146176  601ac142be37ce2ead43755d   
3       0.0  TEST BRANDCODE @1612366146051  601ac142be37ce2ead43755a   
4       0.0  TEST BRANDCODE @1612366146827  601ac142be37ce2ead43755e   

                     cpg_id cpg_ref  
0  601ac114be37ce2ead437550    Cogs  
1  5332f5fbe4b03c9a25efd0ba    Cogs  
2  601ac142be37ce2ead437559    Cogs  
3  601ac142be37ce2ead437559    Cogs  
4  5332fa12e4b03c9a25efd1e7    Cogs

## 1. Design a ER diagram


<img src="ER_diagram.png" width="1000"/>

Note:
- I design the relationship between items and brands as many-to-many, because I noticed there are some barcodes that are associated with multiple brand_id. I would consider changing it to many-to-one after confirming with stakeholders if this is reasonable.

## 2. Write a query (MySQL)
What are the top 5 brands by receipts scanned for most recent month?

In [304]:
# SELECT 
#     b.brand_id,
#     COUNT(DISTINCT r.receipt_id) AS receipt_count
# FROM items as i
# LEFT JOIN brands as b
# ON i.barcode = b.barcode
# LEFT JOIN receipts as r
# ON i.receipt_id = r.receipt_id
# WHERE r.dateScanned BETWEEN DATE_FORMAT(CURDATE() - INTERVAL 1 MONTH, '%Y-%m-01 00:00:00') 
#     AND DATE_FORMAT(LAST_DAY(CURDATE() - INTERVAL 1 MONTH, '%Y-%m-%d 23:59:59'))
# GROUP BY brand_id
# ORDER BY receipt_count DESC
# LIMIT 5

## 3. Evaluate data quality

I focus only on the receipts date, but the same logic could be applied to all tables. I checked the following aspects:
- Uniqueness of receipt_id
- Whether the values in the numeric columns make sense.
- Whether the dates make sense
- Whether the missing values are expected/reasonable

### Duplicates

In [305]:
# check if there are duplicate receipts
receipts['receipt_id'].nunique() == receipts.shape[0]

True

### Value range for numeric column

Here I check if the values in the numeric columns make sense. I will confirm them using business knowledge, e.g. if there is a max for pointsEarned per receipt.

In [306]:
# summary statistics for each numeric column
receipts.describe()

bonusPointsEarned  pointsEarned  purchasedItemCount   totalSpent
count         544.000000    609.000000           635.00000   684.000000
mean          238.893382    585.962890            14.75748    77.796857
std           299.091731   1357.166947            61.13424   347.110349
min             5.000000      0.000000             0.00000     0.000000
25%             5.000000      5.000000             1.00000     1.000000
50%            45.000000    150.000000             2.00000    18.200000
75%           500.000000    750.000000             5.00000    34.960000
max           750.000000  10199.800000           689.00000  4721.950000

In [307]:
# percentage of FINISHED receipts with high pointsEarned
receipts[receipts['rewardsReceiptStatus'] == 'FINISHED']['pointsEarned'].quantile(0.98)

4724.369999999988

In [308]:
# percentage of FINISHED receipts with high values
receipts[receipts['rewardsReceiptStatus'] == 'FINISHED']['purchasedItemCount'].quantile(0.98)

209.9799999999999

### Date range for date column

In [309]:
# check date range for each date column
for col in ['createDate', 'dateScanned', 'finishedDate', 'modifyDate', 'pointsAwardedDate', 'purchaseDate']:
    print(col, receipts[col].min(), receipts[col].max())

createDate 2020-10-30 20:17:59 2021-03-01 23:17:34.772000
dateScanned 2020-10-30 20:17:59 2021-03-01 23:17:34.772000
finishedDate 2021-01-03 15:24:10 2021-02-26 22:36:25
modifyDate 2021-01-03 15:24:10 2021-03-01 23:17:34.772000
pointsAwardedDate 2020-10-30 20:18:00 2021-02-26 22:36:25
purchaseDate 2017-10-30 00:00:00 2021-03-08 17:37:13


##### between createDate and dateScanned

In [310]:
# check if the difference between createDate and dateScanned is reasonable
(receipts['dateScanned'] - receipts['createDate']).describe()

count                         1119
mean     0 days 00:00:00.000000893
std      0 days 00:00:00.000029894
min                0 days 00:00:00
25%                0 days 00:00:00
50%                0 days 00:00:00
75%                0 days 00:00:00
max         0 days 00:00:00.001000
dtype: object

##### between dateScanned and purchaseDate

In [311]:
# check if the difference between dateScanned and purchaseDate is reasonable
(receipts['dateScanned'] - receipts['purchaseDate']).describe()

count                            671
mean      30 days 12:47:14.233709389
std      144 days 18:37:53.408942698
min               -32 days +23:59:59
25%           0 days 19:53:43.500000
50%                  1 days 00:00:00
75%                  3 days 09:06:06
max               1191 days 00:28:36
dtype: object

There are 2 issues I noticed. 
- There are receipts with purchaseDate after dateScanned.
- There are receipts that are at most 1191 days old. I would want to verify if 1) points are reward to these receipts, and 2) if so, is it expected?

In [312]:
# look into the receipts with purchaseDate after dateScanned
sus_receipts = receipts[receipts['purchaseDate'] > receipts['dateScanned']]
print(sus_receipts.shape)
print(sus_receipts['rewardsReceiptStatus'])
sus_receipts.head()


(13, 15)
12      FLAGGED
14     FINISHED
85     FINISHED
139     FLAGGED
158     FLAGGED
190     FLAGGED
244    FINISHED
265     FLAGGED
294     FLAGGED
362    FINISHED
553    FINISHED
644    FINISHED
871    FINISHED
Name: rewardsReceiptStatus, dtype: object


bonusPointsEarned                            bonusPointsEarnedReason  \
12               150.0  Receipt number 5 completed, bonus point schedu...   
14               300.0  Receipt number 4 completed, bonus point schedu...   
85                25.0                        COMPLETE_NONPARTNER_RECEIPT   
139                NaN                                                NaN   
158              500.0  Receipt number 2 completed, bonus point schedu...   

             createDate         dateScanned        finishedDate  \
12  2021-01-03 15:24:38 2021-01-03 15:24:38                 NaT   
14  2021-01-03 15:24:34 2021-01-03 15:24:34 2021-01-03 15:24:35   
85  2021-01-05 20:39:00 2021-01-05 20:39:00 2021-01-05 20:39:00   
139 2021-01-07 16:50:41 2021-01-07 16:50:41                 NaT   
158 2021-01-08 15:02:09 2021-01-08 15:02:09                 NaT   

             modifyDate   pointsAwardedDate  pointsEarned        purchaseDate  \
12  2021-01-03 15:24:38                 NaT        8850.0 2021-02-03 15:24:38   
14  2021-01-03 15:24:35 2021-01-03 15:24:35         300.0 2021-02-03 15:24:35   
85  2021-01-05 20:39:00 2021-01-05 20:39:00          25.0 2021-02-05 20:39:00   
139 2021-01-07 16:50:41                 NaT        8700.0 2021-02-07 16:50:41   
158 2021-01-08 15:02:10                 NaT        9200.0 2021-02-08 15:02:10   

     purchasedItemCount                             rewardsReceiptItemList  \
12                 10.0  [{'barcode': '034100573065', 'description': 'M...   
14                  1.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
85                  1.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
139                10.0  [{'barcode': '034100573065', 'description': 'M...   
158                10.0  [{'barcode': '034100573065', 'description': 'M...   

    rewardsReceiptStatus  totalSpent                    userId  \
12               FLAGGED       290.0  5ff1e194b6a9d73a3a9f1052   
14              FINISHED         1.0  5ff1e194b6a9d73a3a9f1052   
85              FINISHED         1.0  5ff4ce33c3d63511e2a484b6   
139              FLAGGED       290.0  5ff73b90eb7c7d31ca8a452b   
158              FLAGGED       290.0  5ff873d1b3348b11c9337716   

                   receipt_id  
12   5ff1e1b60a7214ada100055c  
14   5ff1e1b20a7214ada100055a  
85   5ff4ce640a7214ada10005e0  
139  5ff73be10a7214ada1000619  
158  5ff873f10a720f052300064f

In [313]:
13 / receipts[(receipts['rewardsReceiptStatus'] == 'FINISHED') | (receipts['rewardsReceiptStatus'] == 'FLAGGED')].shape[0]

0.02304964539007092

**Potential data issue #1: There are 13 'FINISHED' or 'FLAGGED' receipts (~2%) with purchaseDate after dateScanned.**

In [314]:
# look into receipts with large difference between dateScanned and purchaseDate
sus_receipts = receipts[((receipts['dateScanned'] - receipts['purchaseDate']).dt.days > 14) & (receipts['rewardsReceiptStatus'] == 'FINISHED')]
sus_receipts['dateScanned'] - sus_receipts['purchaseDate']

30    15 days 19:48:19
34    16 days 00:00:00
50    16 days 00:00:00
90    16 days 00:00:00
106   15 days 15:06:28
324   16 days 20:27:09
349   16 days 00:00:00
385   15 days 17:03:05
399   15 days 13:58:26
456   16 days 00:00:00
460   16 days 00:00:00
466   16 days 00:00:00
487   15 days 06:09:01
554   16 days 00:00:00
dtype: timedelta64[ns]

In [315]:
sus_receipts['purchaseDate'].min(), sus_receipts['purchaseDate'].max()

(Timestamp('2020-12-19 19:20:35'), Timestamp('2021-01-13 17:50:38'))

**Potential data issue #2: There are 14 receipts older than 14 days accepted (~3% of 'FINISHED' receipts). All happened between 12/19/20 and 1/13/21. Although it's not a common issue, I would like to confirm if it's expected, so that it wouldn't cause problems in the future**

##### between dateScanned and finishedDate

In [316]:
# check if the difference between dateScanned and finishedDate is reasonable
(receipts['dateScanned'] - receipts['finishedDate']).describe()

count                            568
mean     -1 days +11:50:57.695790493
std        5 days 15:16:38.187017472
min               -79 days +04:08:36
25%                -1 days +23:59:58
50%                -1 days +23:59:59
75%                  0 days 00:00:00
max                  0 days 00:00:00
dtype: object

##### between finishedDate and pointsAwardedDate

In [317]:
# check if the difference between pointsAwardedDate and finishedDate is reasonable
(receipts['pointsAwardedDate'] - receipts['finishedDate']).describe()

count                            514
mean     -1 days +23:55:36.964980545
std        0 days 01:07:19.056138505
min                -1 days +05:59:53
25%                  0 days 00:00:00
50%                  0 days 00:00:00
75%                  0 days 00:00:00
max                  0 days 00:00:00
dtype: object

### Missing values

Here I look into whether missing values are because we are still processing the receipts or because we fail to collect the data.

In [318]:
# percentage of missing values in each column, by rewardsReceiptStatus
receipts.groupby('rewardsReceiptStatus').apply(lambda x: x.isnull().sum() / x.shape[0]).transpose()

rewardsReceiptStatus     FINISHED   FLAGGED  PENDING  REJECTED  SUBMITTED
bonusPointsEarned        0.119691  0.347826     1.00  0.183099        1.0
bonusPointsEarnedReason  0.119691  0.347826     1.00  0.183099        1.0
createDate               0.000000  0.000000     0.00  0.000000        0.0
dateScanned              0.000000  0.000000     0.00  0.000000        0.0
finishedDate             0.000000  1.000000     0.00  1.000000        1.0
modifyDate               0.000000  0.000000     0.00  0.000000        0.0
pointsAwardedDate        0.007722  0.586957     1.00  0.943662        1.0
pointsEarned             0.000000  0.282609     1.00  0.183099        1.0
purchaseDate             0.000000  0.239130     0.02  0.028169        1.0
purchasedItemCount       0.000000  0.000000     1.00  0.000000        1.0
rewardsReceiptItemList   0.003861  0.000000     0.02  0.042254        1.0
rewardsReceiptStatus     0.000000  0.000000     0.00  0.000000        0.0
totalSpent               0.000000  0.000000     0.02  0.000000        1.0
userId                   0.000000  0.000000     0.00  0.000000        0.0
receipt_id               0.000000  0.000000     0.00  0.000000        0.0

There are several interesting findings. If time allows, I would look into and confirm with stakeholders if they are expected (if it's of business interest).
- Some 'REJECTED' receipts still seem to have received points.
- Although very rare, there are some 'FINISHED' receipts with missing item list and/or pointsAwardedDate
- There are a lot of missing values in 'FLAGGED' receipts. I would like to know what this status mean and if these missing values make sense.



Additionally, I want to quickly examine the percentage of missing in rewardsReceiptItemList column.

In [319]:
# percentage of missing values in items
items.apply(lambda x: x.isnull().sum() / x.shape[0]).transpose()

barcode                               0.554819
description                           0.054891
finalPrice                            0.025068
itemPrice                             0.025068
needsFetchReview                      0.882870
partnerItemId                         0.000000
preventTargetGapPoints                0.948422
quantityPurchased                     0.025068
userFlaggedBarcode                    0.951448
userFlaggedNewItem                    0.953465
userFlaggedPrice                      0.956923
userFlaggedQuantity                   0.956923
receipt_id                            0.000000
receipt_rewardsReceiptStatus          0.000000
needsFetchReviewReason                0.968448
pointsNotAwardedReason                0.951016
pointsPayerId                         0.817461
rewardsGroup                          0.750612
rewardsProductPartnerId               0.673102
userFlaggedDescription                0.970465
originalMetaBriteBarcode              0.989771
originalMetaB

In [320]:
# percentage of missing values in items, by rewardsReceiptStatus
items.groupby('receipt_rewardsReceiptStatus').apply(lambda x: x.isnull().sum() / x.shape[0]).transpose()

receipt_rewardsReceiptStatus        FINISHED   FLAGGED   PENDING  REJECTED
barcode                             0.535485  0.703704  1.000000  0.384146
description                         0.034133  0.155556  0.000000  0.323171
finalPrice                          0.004055  0.130864  0.000000  0.268293
itemPrice                           0.004055  0.130864  0.000000  0.268293
needsFetchReview                    0.891686  0.848148  1.000000  0.701220
partnerItemId                       0.000000  0.000000  0.000000  0.000000
preventTargetGapPoints              0.968570  0.848148  1.000000  0.701220
quantityPurchased                   0.004055  0.130864  0.000000  0.268293
userFlaggedBarcode                  0.972119  0.848148  1.000000  0.701220
userFlaggedNewItem                  0.974485  0.848148  1.000000  0.701220
userFlaggedPrice                    0.978540  0.848148  1.000000  0.701220
userFlaggedQuantity                 0.978540  0.848148  1.000000  0.701220
receipt_id                          0.000000  0.000000  0.000000  0.000000
receipt_rewardsReceiptStatus        0.000000  0.000000  0.000000  0.000000
needsFetchReviewReason              0.991382  0.848148  1.000000  0.725610
pointsNotAwardedReason              0.942548  1.000000  1.000000  1.000000
pointsPayerId                       0.798074  0.911111  1.000000  1.000000
rewardsGroup                        0.735046  0.804938  1.000000  0.969512
rewardsProductPartnerId             0.632984  0.886420  1.000000  0.969512
userFlaggedDescription              0.992734  0.855556  1.000000  0.725610
originalMetaBriteBarcode            0.988003  1.000000  1.000000  1.000000
originalMetaBriteDescription        0.998310  1.000000  1.000000  1.000000
brandCode                           0.602906  0.693827  0.979592  0.993902
competitorRewardsGroup              0.960460  0.950617  0.979592  1.000000
discountedItemPrice                 0.150389  0.260494  0.000000  0.432927
originalReceiptItemText             0.152247  0.258025  0.000000  0.432927
itemNumber                          0.974147  1.000000  1.000000  1.000000
originalMetaBriteQuantityPurchased  0.997465  1.000000  1.000000  1.000000
pointsEarned                        0.855526  0.911111  1.000000  1.000000
targetPrice                         0.946266  0.925926  1.000000  1.000000
competitiveProduct                  0.895235  0.975309  1.000000  0.969512
originalFinalPrice                  0.998479  1.000000  1.000000  1.000000
originalMetaBriteItemPrice          0.998479  1.000000  1.000000  1.000000
deleted                             1.000000  1.000000  1.000000  0.945122
priceAfterCoupon                    0.861609  1.000000  0.020408  0.457317
metabriteCampaignId                 0.865495  0.917284  1.000000  1.000000

There is one particularly interesting finding -- More than half of the items in 'FINISHED' receipts don't have a barcode. This is problematic since we are joining the brands table with items table using this column. Thus, our answers to questions like 'what are the top 5 brands last month' might be incorrect. It's worth flagging this issue to stakeholders to see why this is the case and if there is a workaround.


**Potential data issue #3: 50%+ receipts have missing barcodes, which might lead to a biased understanding of brands distribution.**

In [321]:
brands['barcode'] = brands['barcode'].astype(str)
items.merge(brands, on='barcode', how='left')['brand_id'].isnull().sum() / items.shape[0]

0.9881861403256015

In [322]:
# check if a barcode is associated with multiple brands
brands.groupby('barcode')['brand_id'].nunique().sort_values(ascending=False).head(10)

barcode
511111504139    2
511111305125    2
511111704140    2
511111605058    2
511111004790    2
511111504788    2
511111204923    2
511111000167    1
511111616184    1
511111616535    1
Name: brand_id, dtype: int64

In [323]:
7 / brands['barcode'].nunique()

0.00603448275862069

With current provided data, 99% of the items in the receipts don't have a corresponding brand when using barcode to join the items and brands table. I think this is because 1) brands data is not comprehensive, and 2) 50%+ barcode are missing in the receipts. I would like to know if there is a more comprehensive list of the brands to estimate the effects of missing barcode.

### 4. Communicate with stakeholders

Hi all,

I'm reaching out regarding the project on designing a database to store our receipts, users, and brands information. I have examined the historical data, and I have a few items that need clarification and support from you to finalize the design.

- **Major issues - these are blockers of the project**
1. To understand popularity and user behaviors for different brands, I'm using barcode to associate items in receipts and their brands. There are 2 concerns with this. 
   1. Given current data I could only match 1% of the items with their brands. Could you provide a more comprehensive list of the brands, so that I could test the feasibility of this approach? 
   2. 50%+ of the items have missing barcode. For these items, is there another way to extract their brand information? Otherwise, our analysis of brands distribution could be biased.
2. The size of the items table might grow quickly in the future. There are several ways we could ensure scalability. I would love your thoughts on which one we should proceed with. 
   1. Condense less used columns (e.g originalMetaBrite... and userFlagged... columns) into json fields -- simplest to implement but least effective
   2. Store columns that have 1-to-1 matching (e.g. item-specific columns like original price, description, pointsPayerId, rewardsGroup) to a separate table -- applicable only if we have 1-to-1 matching relationships
   3. Use other non-relational databases like NoSQL, which is better at handling json files and less costly to maintain -- most effective, but require system-level changes 

- **Minor issues - we should look into these if time allows to ensure data quality in the long term**
1. ~2% of accepted receipts have 5k+ points earned (up to ~10k points) and 200+ items (up to ~700 items) purchased. Is this reasonable, or do we have upper limits on these values?
2. ~3% of acepted receipts are older than 14 days, all between 12/19/20 and 1/13/21. From my understanding we only accept receipts that are within 14 days of purchase, so just want to confirm if this is expected.
3. ~2% of accepted or flagged receipts have purchase date after scan date. Is this a system glitch? If so, have we fixed it?

Your help on these items would be very much appreciated. I'd be happy to schedule a meeting to go through them in more details as well. Thanks!

Best,

Qingyang
